### PR TITLE
app-emulation/virtualbox-modules: fix kernel >= 5.18 build fail

### PR DESCRIPTION
**Pull request by peeweep**
Closes: https://bugs.gentoo.org/847094

Signed-off-by: peeweep <peeweep@0x0.ee>

### CODE REVIEW

**Comment by aristocrates** [comment]
In https://www.virtualbox.org/ticket/20914 there is a patch [[1]](https://www.virtualbox.org/raw-attachment/ticket/20914/vbox-linux-5.18.patch) that includes a number of changes  - should the patch include the other relevant changes? Another example: the opensuse patch [[2]](https://build.opensuse.org/package/view_file/Virtualization/virtualbox/fixes_for_kernel_5.18.patch?expand=1)

[1] https://www.virtualbox.org/raw-attachment/ticket/20914/vbox-linux-5.18.patch
* Note that some of the files modified in that patch are not present when `vbox-kernel-module-src-${PV}.tar.gz` is extracted, and others appear to have different full paths in the source tree

[2] https://build.opensuse.org/package/view_file/Virtualization/virtualbox/fixes_for_kernel_5.18.patch?expand=1

**Comment by peeweep** [comment]
@aristocrates Thank you for providing information, I haven't seen this ticket before. I'm not sure about the influence of vbox-linux-5.18.patch (except for fixing the compile problem, I can understand this). I think this PR is best used to fix compile error. Do not introduce other problems.

**Comment by ignamartinoli** [comment]
Will this change target kernel 5.18.9? It would be awesome

File: app-emulation/virtualbox-modules/virtualbox-modules-6.1.34.ebuild [hunk 1]
```diff
@@ -50,6 +50,9 @@ src_prepare() {
 	if use pax-kernel && kernel_is -ge 3 0 0 ; then
 		eapply -p0 "${FILESDIR}"/${PN}-5.2.8-pax-const.patch
 	fi
+	if kernel_is -ge 5 18 0 ; then
```
Comment (Peter-Levine) [inline app-emulation/virtualbox-modules/virtualbox-modules-6.1.34.ebuild:53]: It shouldn't be necessary to check the kernel version in the ebuild as the preprocessor includes a conditional check, i.e.,  `#if RTLNX_VER_MIN(5,18,0)`.  You could just change the patch from
```
--- vboxnetflt/linux/VBoxNetFlt-linux.c
+++ vboxnetflt/linux/VBoxNetFlt-linux.c
```
to
```
--- a/vboxnetflt/linux/VBoxNetFlt-linux.c
+++ b/vboxnetflt/linux/VBoxNetFlt-linux.c
```
and add `PATCHES=( "${FILESDIR}"/${PN}-5.2.8-pax-const.patch" )` in global scope instead.

Comment (peeweep) [reply app-emulation/virtualbox-modules/virtualbox-modules-6.1.34.ebuild]: Thanks, I added ` PATCHES=( "${FILESDIR}"/${PN}-6.1.34-remove-netif_rx_ni.patch )`, it looks better.

**Comment by M1cR0xf7** [comment]
any updates? is this going to get merged

**Comment by xfzv** [comment]
> any updates? is this going to get merged

VirtualBox 6.1.36 was released on July 19th

From [changelog](https://www.virtualbox.org/wiki/Changelog-6.1#v36):

```
- Linux Host and Guest drivers: Introduced initial support for kernels 5.18, 5.19 and RHEL 9.1 (bugs [#20914](https://www.virtualbox.org/ticket/20914), [#20941](https://www.virtualbox.org/ticket/20941))
```

So I guess the patch won't be required anymore?

**Comment by ignamartinoli** [comment]
@xfzv I think maybe it's not applied?
I just tried to emerge it and

```
sicro@sicro ~ $ doas emerge --sync
>>> Syncing repository 'gentoo' into '/var/db/repos/gentoo'...
/var/db/repos/gentoo
/usr/bin/git fetch origin
 * Using keys from /usr/share/openpgp-keys/gentoo-release.asc
 * Refreshing keys via WKD ...                                                                                                                                                                                                           [ ok ]
 * Trusted signature found on top commit
Already up to date.
=== Sync completed for gentoo
>>> Syncing repository 'guru' into '/var/db/repos/guru'...
/var/db/repos/guru
/usr/bin/git fetch origin
Already up to date.
=== Sync completed for guru

Action: sync for repo: gentoo, returned code = 0
Action: sync for repo: guru, returned code = 0


sicro@sicro ~ $ doas emerge virtualbox
Calculating dependencies... done!
>>> Verifying ebuild manifests
>>> Running pre-merge checks for app-emulation/virtualbox-6.1.34-r6
 * You have disabled the "python" USE flag. This will only
 * disable the python bindings being installed.
>>> Emerging (1 of 9) app-emulation/virtualbox-modules-6.1.34::gentoo
>>> Failed to emerge app-emulation/virtualbox-modules-6.1.34, Log file:
>>>  '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/temp/build.log'
>>> Jobs: 0 of 9 complete, 1 failed                 Load avg: 2.58, 1.82, 1.55
 * Package:    app-emulation/virtualbox-modules-6.1.34
 * Repository: gentoo
 * Maintainer: ceamac.paragon@gmail.com proxy-maint@gentoo.org
 * USE:        abi_x86_64 amd64 elibc_glibc kernel_linux userland_GNU
 * FEATURES:   ccache network-sandbox preserve-libs sandbox userpriv usersandbox
 * Determining the location of the kernel source code
 * Found kernel source directory:
 *     /usr/src/linux
 * Found sources for kernel version:
 *     5.18.12-gentoo
 * Checking for suitable kernel configuration options ...
 [ ok ]
>>> Unpacking source...
>>> Unpacking vbox-kernel-module-src-6.1.34.tar.xz to /var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work
>>> Source unpacked in /var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work
>>> Preparing source in /var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work ...
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work ...
>>> Source configured.
>>> Compiling source in /var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work ...
ln: failed to create symbolic link 'Module.symvers': File exists
 * Preparing vboxdrv module

... Infine logs ...

make[2]: Leaving directory '/usr/src/linux-5.18.12-gentoo'
make[1]: Leaving directory '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work/vboxnetadp'

/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work/vboxnetflt/linux/VBoxNetFlt-linux.c: In function ‘vboxNetFltPortOsXmit’:
/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work/vboxnetflt/linux/VBoxNetFlt-linux.c:2314:23: error: implicit declaration of function ‘netif_rx_ni’; did you mean ‘netif_rx’? [-Werror=implicit-function-declaration]
 2314 |                 err = netif_rx_ni(pBuf);
      |                       ^~~~~~~~~~~
      |                       netif_rx
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:288: /var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work/vboxnetflt/linux/VBoxNetFlt-linux.o] Error 1
make[2]: *** [Makefile:1838: /var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work/vboxnetflt] Error 2
make[2]: Leaving directory '/usr/src/linux-5.18.12-gentoo'
make[1]: *** [/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work/vboxnetflt/Makefile-footer.gmk:117: vboxnetflt] Error 2
make[1]: Leaving directory '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work/vboxnetflt'
make: *** [Makefile:65: vboxnetflt] Error 1
 * ERROR: app-emulation/virtualbox-modules-6.1.34::gentoo failed (compile phase):
 *   emake failed
 *
 * If you need support, post the output of `emerge --info '=app-emulation/virtualbox-modules-6.1.34::gentoo'`,
 * the complete build log and the output of `emerge -pqv '=app-emulation/virtualbox-modules-6.1.34::gentoo'`.
 * The complete build log is located at '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/temp/environment'.
 * Working directory: '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work'
 * S: '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work'

 * Messages for package app-emulation/virtualbox-modules-6.1.34:

 * ERROR: app-emulation/virtualbox-modules-6.1.34::gentoo failed (compile phase):
 *   emake failed
 *
 * If you need support, post the output of `emerge --info '=app-emulation/virtualbox-modules-6.1.34::gentoo'`,
 * the complete build log and the output of `emerge -pqv '=app-emulation/virtualbox-modules-6.1.34::gentoo'`.
 * The complete build log is located at '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/temp/environment'.
 * Working directory: '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work'
 * S: '/var/tmp/portage/app-emulation/virtualbox-modules-6.1.34/work'
```

It still complains about `netif_rx_ni`

**Comment by ignamartinoli** [comment]
@xfzv so we should just upsource **VirtualBox**?

**Comment by xfzv** [comment]
> @xfzv so we should just upsource **VirtualBox**?

Arch Linux maintainer [deleted the patch](https://github.com/archlinux/svntogit-community/commit/63edde0a20d897686008f26d449d378ac02b0c2f#diff-1aa8d9105b0f86f925027ce20b885eb6e21c919261367e23b3372e9fa5521cf9L246-L262) (click on `Load diff` on `trunk/020-linux518.patch` to see relevant lines) when updating PKGBUILD to 6.1.36, that's why I assume it will no longer be required on Gentoo either.

**Comment by peeweep** [comment]
I think the bug can be closed when Gentoo's virtualbox is upgraded to 6.1.36

**Comment by ignamartinoli** [comment]
@peeweep agreed, is there a way to ask for this to the maintainers?

**Comment by ignamartinoli** [comment]
Hi, I'm running on testing as my `make.conf` says

```sh
ACCEPT_KEYWORDS="~amd64"
```

I did `emerge --sync` and when I ran `emerge virtualbox -a` I got

```
These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild  N     ] app-emulation/virtualbox-modules-6.1.34  USE="-dist-kernel -pax-kernel"
[ebuild  N     ] dev-libs/libIDL-0.8.14-r3
[ebuild  N     ] dev-qt/qtprintsupport-5.15.5  USE="cups -debug -gles2-only -test"
[ebuild  N     ] dev-qt/qtopengl-5.15.5  USE="-debug -gles2-only -test"
[ebuild  N     ] dev-qt/linguist-tools-5.15.5  USE="-debug -qml -test"
[ebuild  N     ] dev-util/kbuild-0.1.9998.3499
[ebuild  N     ] media-libs/audiofile-0.3.6-r4  USE="-flac"
[ebuild  N     ] media-libs/libsdl-1.2.15_p20210224  USE="X alsa joystick opengl pulseaudio sound video -aalib -custom-cflags -dga -fbcon -libcaca -nas -oss -static-libs -tslib -xinerama -xv"
[ebuild  N     ] app-emulation/virtualbox-6.1.34-r6  USE="alsa opengl opus pam pulseaudio qt5 sdk udev -debug -doc -dtrace -headless -java -lvm -pax-kernel (-pch) -python -vboxwebsrv -vnc" PYTHON_SINGLE_TARGET="python3_10 -python3_8 -python3_9"
```

even though `virtualbox` version was updated, as shown in https://packages.gentoo.org/packages/app-emulation/virtualbox.
Is it normal? I wanted to test the update

**Comment by thesamesam** [comment]
Obsolete given bump is done.

**Pull request closed without merging**